### PR TITLE
[Feat]: Gem builder documentation

### PR DESCRIPTION
**Pull request by poojaSprophecy**

### CODE REVIEW

File: docs/low-code-spark/gems/source-target/random-data-creator.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+---
+title: Random Data Generator
```
Comment (slyons) [inline docs/low-code-spark/gems/source-target/random-data-creator.md:2]: Can this file be moved into another PR? It requires a bit more work.

File: docs/low-code-spark/extensibility/gem-ui-componets.md [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+openCreateDatasetDialog(state.asInstanceOf[WorkflowProcess])
+}
+)
+```
+
+</TabItem>
+</Tabs>
+````
+
+### Forms and Capturing Data
```
Comment (slyons) [inline docs/low-code-spark/extensibility/gem-ui-componets.md:329]: We can't leave unfinished sections like this

**Review by slyons — CHANGES_REQUESTED** [review]

File: docs/low-code-spark/gems/source-target/random-data-creator.md (hunk 1 above)
Comment (poojaSprophecy) [reply docs/low-code-spark/gems/source-target/random-data-creator.md:2]: yes. Removed from this PR.

File: docs/low-code-spark/extensibility/gem-ui-componets.md (hunk 2 above)
Comment (poojaSprophecy) [reply docs/low-code-spark/extensibility/gem-ui-componets.md]: Done.

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 3]
```diff
@@ -6,20 +6,859 @@ description: Gem-Builder
...
+</TabItem>
+</Tabs>
+
+````
+
+You can go ahead and preview the component in the Gem Builder now to see how it looks like. You can modify the properties and then save it to preview the generated Spark code which will eventually run on your cluster.
+
+## DataSources Gems
+
+DataSource Gems (also referred to as Source/Target Gems) are Gems which you use to read/write your Datasets into DataFrames. The only major difference between a Source/Target Gem and a Transformation Gem is that a Source/Target Gem will have two `dialog` and two `appl`y functions each for Source and Target respectively. Let's look at them with an example
```
Comment (ankitashukla) [inline docs/low-code-spark/extensibility/gem-builder.md:438]: minor typo: and two `apply` functions

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 4]
```diff
@@ -6,20 +6,859 @@ description: Gem-Builder
...
+````
+
+### Parent Class
+
+Your Gem class needs to extend a parent class from which it inherits the representation of the overall Gem. This includes the UI and the logic.
+For transform Gems you need to extend `ComponentSpec` (Like in the example above), for Source/Target Gem you need to extend `DatasetSpec`. We will see the difference between the two at the end.
+
+First thing you give after this is the name and category of your Gem, `"Filter"` and `"Transform"` in this example.
+
+Another thing note here is `optimizeCode`. This function returns a `True` or `False` value depending on whether we want the Prophecy Optimizer to run on this code to simplify it.
```
Comment (ankitashukla) [inline docs/low-code-spark/extensibility/gem-builder.md:195]: ...here is `optimizeCode`. This flag can be set to `True` or `False`...

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 5]
```diff
@@ -6,20 +6,859 @@ description: Gem-Builder
...
+    }
+
+  }
+```
+</TabItem>
+</Tabs>
+
+````
+
+You can go ahead and preview the component in the Gem Builder now to see how it looks like. You can modify the properties and then save it to preview the generated Spark code which will eventually run on your cluster.
```
Comment (ankitashukla) [inline docs/low-code-spark/extensibility/gem-builder.md:434]: to see how it looks.

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 6]
```diff
@@ -6,20 +6,859 @@ description: Gem-Builder
...
+</TabItem>
+</Tabs>
+
+````
+
+Let's look at each of these functions.
+
+### Dialog
+
+The `dialog` function contain code specific to how the Gem UI should look to the user.
```
Comment (ankitashukla) [inline docs/low-code-spark/extensibility/gem-builder.md:267]: function contains

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 7]
```diff
@@ -6,20 +6,859 @@ description: Gem-Builder
...
 </div>
-
 <br />
 
-:::caution Learn more
+## Getting Started
+
+All of our Gems are stored at account level. Once you navigate to the Gem listing page you will see all the Prophecy defined and User Defined Gems. You can add a new Gem or modify an existing one.
+
+After writing your Gem code, you can click `Preview` to see, how the Gem UI look. You can fill in some values, and click save to check the Spark code which will be generated for the Gem.
```
Comment (ankitashukla) [inline docs/low-code-spark/extensibility/gem-builder.md:30]: see how the Gem UI looks.

File: docs/low-code-spark/extensibility/gem-builder.md (hunk 4 above)
Comment (ankitashukla) [inline docs/low-code-spark/extensibility/gem-builder.md:195]: Another thing to* note here

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 8]
```diff
@@ -6,20 +6,859 @@ description: Gem-Builder
...
+```
+
+</TabItem>
+</Tabs>
+````
+
+### Parent Class
+
+Your Gem class needs to extend a parent class from which it inherits the representation of the overall Gem. This includes the UI and the logic.
+For transform Gems you need to extend `ComponentSpec` (Like in the example above), for Source/Target Gem you need to extend `DatasetSpec`. We will see the difference between the two at the end.
```
Comment (ankitashukla) [inline docs/low-code-spark/extensibility/gem-builder.md:191]: For Transform Gems, you need to extend `ComponentSpec` (like* in the ..

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 9]
```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
...
-Prophecy Data Engineering system can be extended with **Gem Builder** to add **your own standard sources, targets and
-transforms** to Spark and roll it out to your entire team. You can build **your own frameworks with** custom Data Quality
-library, or auditing library.
+:::caution Enterprise Only
+
+Please [contact us](https://www.prophecy.io/request-a-demo) to learn more about the Enterprise offering.
+
+:::
+
+[Gems](docs/concepts/gems.md) are operations you can add to your Pipeline to perform actions on your data. The **Gem Builder** allows you to create custom UI experiences that will generate optimized code wherever it's used. You can add your own custom sources, targets, and transformations and roll them out to your entire team. You can even build a custom Data Quality library, or an auditing library.
```
Comment (atbida) [inline docs/low-code-spark/extensibility/gem-builder.md:16]: Each Prophecy Pipeline is composed of individual operations, or [Gems](docs/concepts/gems.md), that perform actions on data. While Prophecy offers dozens of Gems out-of-the-box, some data practitioners want to extend this idea and create their own Gems. Luckily, **Gem Builder** allows enterprise users to add custom Gems. Create custom source, target, and transformation Gems, publish, and your team can utilize your custom Gem. Like all Prophecy Gems, the Custom Gem code will be saved to git using Python or Scala. 

Consider repeated patterns when deciding when to create a custom Gem. Does your team use a particular business logic frequently? Do you notice yourself doing the same Data Quality checks more than once a month? Does your team have an auditing use case that could benefit from re-usable code? Consider a custom Gem to capture these logical patterns for re-use!

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 10]
```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
...
 allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen>
 </iframe>
 </div>
-
 <br />
 
-:::caution Learn more
+## Getting Started
+
+All of our Gems are stored at the account level. Once you navigate to the Gem listing page you will see all the Prophecy-defined and User Defined Gems. You can add a new Gem or modify an existing one.
```
Comment (atbida) [inline docs/low-code-spark/extensibility/gem-builder.md:26]: Custom Gem logic can be shared with other users within the Team and Organization. 

Pooja - I'm not sure what is meant by "Gem listing page"

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 11]
```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
...
+
+Please [contact us](https://www.prophecy.io/request-a-demo) to learn more about the Enterprise offering.
+
+:::
+
+[Gems](docs/concepts/gems.md) are operations you can add to your Pipeline to perform actions on your data. The **Gem Builder** allows you to create custom UI experiences that will generate optimized code wherever it's used. You can add your own custom sources, targets, and transformations and roll them out to your entire team. You can even build a custom Data Quality library, or an auditing library.
 
 <div class="video-container">
 <iframe src="https://www.youtube.com/embed/K23pOatAeVE" title="YouTube video player" frameborder="0"
 allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen>
```
Comment (atbida) [inline docs/low-code-spark/extensibility/gem-builder.md:19]: This video should be re-created with the new UI. 
The same audio could be used :-)

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 12]
```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
...
+After writing your Gem code, you can click `Preview` to see how the Gem UI looks. You can fill in some values, and click save to check the Spark code which will be generated for the Gem.
+Finally, just publish the Gems and your Gem is ready to be used in the Pipelines.
+
+Please refer below video for a step-by-step example:
+
+<div class="wistia_responsive_padding" style={{padding:'56.25% 0 0 0', position:'relative'}}>
+<div class="wistia_responsive_wrapper" style={{height:'100%',left:0,position:'absolute',top:0,width:'100%'}}>
+<iframe src="https://user-images.githubusercontent.com/121796483/212867119-a05b1204-0be7-4f0b-866b-675b4b0b4280.mp4" title="Gem builder" allow="autoplay;fullscreen" allowtransparency="true" frameborder="0" scrolling="no" class="wistia_embed" name="wistia_embed" msallowfullscreen width="100%" height="100%"></iframe>
+</div></div>
+
```
Comment (atbida) [inline docs/low-code-spark/extensibility/gem-builder.md:37]: In this video, I suggest going a bit slower to show that the code is being copied from some text editor to the Prophecy UI.

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 13]
```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
...
+## Tutorial
+
+The Gem builder is a tool that enables users to create any custom Gems or modify existing ones. There are two types of Gems:
+
+- **DataSource Gems**: These Gems enable the reading and writing of data from or to various data sources
+- **Transform Gems**: These Gems apply transformations/joins/any other custom logic onto any DataFrame(s) that are passed into them.
```
Comment (atbida) [inline docs/low-code-spark/extensibility/gem-builder.md:43]: The above text doesn't match the UI: 
image


According to this UI, it would appear that I can create a Gem.Gem or a Gem.DataSource. 
Furthermore, either Gem.Gem or Gem.DataSource could be a Custom, Transform, Join/Split, MyGem, or Hidden. 

The text is much clearer than the UI...can the UI be clarified?
Perhaps the user needs to select 
Create custom Gem!
Choose: should this gem be shared? If so, select teams (or projects?) that should be able to use this Custom Gem.
Choose: DataSource, DataTarget, or Transformation
If Transformation, 
Choose: Join/Split or Other

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 14]
```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
...
+
+Programmatically, a Gem is a component with the following parts:
+
+- The **Gem UI Component** to get user information from the screen (This code is rendered on the Prophecy UI)
+- The **Gem Code Logic** which is how the Gem acts within the context of a Pipeline.
+
+You can write Gem code using either Python or Scala
```
Comment (atbida) [inline docs/low-code-spark/extensibility/gem-builder.md:50]: Gem code can be written using either Python or Scala.

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 15]
```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
...
+
+## Defining a Gem
+
+### Example
+
```
Comment (atbida) [inline docs/low-code-spark/extensibility/gem-builder.md:55]: @poojaSprophecy could you pls add comments to this code snippet?

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 16]
```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
...
+The above Dialog code in the filter is rendered on UI like this
+![Dialog](img/gem-builder-ui.png)
+
+There are various UI components that can be defined for your custom Gems such as scroll boxes, tabs, buttons, and more! These UI components can be grouped together in various types of panels to create a custom user experience when using the Gem.
+
+After the Dialog object is defined, it's serialized as JSON, sent to the UI, and rendered there.
 
-Please [contact us](https://www.prophecy.io/request-a-demo) to learn more about the Gem builder.
+Depending on what kind of Gem you're creating, you'll have to define either a `Dialog` or a `DatasetDialog`.
```
Comment (atbida) [inline docs/low-code-spark/extensibility/gem-builder.md:318]: either a `Transformation Dialog` or a `Dataset Dialog`

File: docs/low-code-spark/extensibility/gem-builder.md (hunk 13 above)
Comment (atbida) [reply docs/low-code-spark/extensibility/gem-builder.md:43]: Edit: below the DataSource is called a `Dataset` ... so we need to make this consistent.

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 17]
```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
...
+
+````
+
+### Properties Classes
+
+There is one class (seen here as `FilterProperties`) that contains a list of the properties to be made available to the user for this particular Gem. Think of these as all the values a user fills out within the template of this Gem, or any other UI state that you need to maintain (seen here as `columnsSelector` and `condition`).
+
+:::caution
+
+The content of these `Properties` classes is persisted in JSON and stored in Git, so be careful not to keep any information in them that could be a potential security risk.
```
Comment (atbida) [inline docs/low-code-spark/extensibility/gem-builder.md:229]: The content of these `Properties` classes is persisted in JSON and stored in Git. 

(omit the remainder of the sentence)

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 18]
```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
...
+</TabItem>
+</Tabs>
+
+````
+
+You can go ahead and preview the component in the Gem Builder now to see how it looks. You can modify the properties and then save it to preview the generated Spark code which will eventually run on your cluster.
+
+## DataSources Gems
+
+DataSource Gems (also referred to as Source/Target Gems) are Gems that you use to read/write your Datasets into DataFrames. The only major difference between a Source/Target Gem and a Transformation Gem is that a Source/Target Gem will have two `dialog` and two `apply` functions each for Source and Target respectively. Let's look at them with an example
```
Comment (atbida) [inline docs/low-code-spark/extensibility/gem-builder.md:436]: In this paragraph there's only one major difference, but at the end we see there are three differences listed. 
I suggest to be consistent - mention all three differences in both the intro and the summary at the end.

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 19]
```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
...
+    class FilterProperties(ComponentProperties):
+        columnsSelector: List[str] = field(default_factory=list)
+        condition: SColumn = SColumn("lit(True)")
+```
+</TabItem>
+
+<TabItem value="scala" label="Scala">
+
+```scala
+object Filter extends ComponentSpec {
```
Comment (pateash) [inline docs/low-code-spark/extensibility/gem-builder.md:251]: @poojaSprophecy, I think we forgot the update scala code here.

File: docs/low-code-spark/extensibility/gem-builder.md [hunk 20]
```diff
@@ -6,20 +6,857 @@ description: Gem-Builder
...
+<Tabs>
+
+<TabItem value="py" label="Python">
+
+```py
+class FilterCode(ComponentCode):
+def __init__(self, newProps):
+self.props: Filter.FilterProperties = newProps
+
+        def apply(self, spark: SparkSession, in0: DataFrame) -> DataFrame:
```
Comment (pateash) [inline docs/low-code-spark/extensibility/gem-builder.md:411]: The indentation for the Python Snippet looks weird.

File: docs/low-code-spark/extensibility/gem-builder.md (hunk 16 above)
Comment (poojaSprophecy) [reply docs/low-code-spark/extensibility/gem-builder.md:318]: This is the name of the actual class we use in code. May be later we can update this if we change class names.

File: docs/low-code-spark/extensibility/gem-builder.md (hunk 11 above)
Comment (poojaSprophecy) [reply docs/low-code-spark/extensibility/gem-builder.md:19]: I will do that in a separate PR may be in few days. :-)

File: docs/low-code-spark/extensibility/gem-builder.md (hunk 13 above)
Comment (poojaSprophecy) [reply docs/low-code-spark/extensibility/gem-builder.md:43]: Updating the screenshot with new UI.

**Review by jainabhinav — APPROVED** [review]
LGTM, let's push out the first draft. We can enhance with more examples later.

**Review by slyons — APPROVED** [review]
